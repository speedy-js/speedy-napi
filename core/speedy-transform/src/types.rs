#[derive(Debug)]
pub struct TransformConfig {
  pub react_runtime: Option<bool>,
  pub babel_import: Option<Vec<BabelImportConfig>>,
}

#[derive(Debug)]
pub struct BabelImportConfig {
  pub from_source: String,
  pub replace_css: Option<RepalceCssConfig>,
  pub replace_js: Option<RepalceSpecConfig>,
}

#[derive(Debug)]
pub struct RepalceSpecConfig {
  pub replace_expr: String,
  pub ignore_es_component: Option<Vec<String>>,
  pub lower: Option<bool>,
}

#[derive(Debug)]
pub struct RepalceCssConfig {
  pub ignore_style_component: Option<Vec<String>>,
  pub replace_expr: String,
  pub lower: Option<bool>,
}
