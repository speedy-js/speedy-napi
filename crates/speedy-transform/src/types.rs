use napi::JsFunction;
use serde::Serialize;

#[napi(object)]
#[derive(Debug)]
pub struct TransformOutput {
  pub code: String,
  pub map: Option<String>,
}

#[napi(object)]
#[derive(Serialize)]
pub struct TransformConfig {
  pub remove_useEffect: Option<bool>,
  pub react_runtime: Option<bool>,
  pub babel_import: Option<Vec<BabelImportConfig>>,
}

#[napi(object)]
#[derive(Serialize)]
pub struct BabelImportConfig {
  pub from_source: String,
  pub replace_css: Option<RepalceCssConfig>,
  pub replace_js: Option<RepalceSpecConfig>,
}

#[napi(object)]
#[derive(Serialize)]
pub struct RepalceSpecConfig {
  #[serde(skip_serializing)]
  pub replace_expr: JsFunction,
  pub ignore_es_component: Option<Vec<String>>,
  pub lower: Option<bool>,
  pub camel2_dash_component_name: Option<bool>,
  pub transform_to_default_import: Option<bool>,
}

#[napi(object)]
#[derive(Serialize)]
pub struct RepalceCssConfig {
  pub ignore_style_component: Option<Vec<String>>,
  #[serde(skip_serializing)]
  pub replace_expr: JsFunction,
  pub lower: Option<bool>,
  pub camel2_dash_component_name: Option<bool>,
}
