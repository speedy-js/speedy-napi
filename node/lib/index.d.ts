import { transformBabelImport } from '../index';

declare const transform: {
  transformBabelImport: typeof transformBabelImport;
};

export { transform };
