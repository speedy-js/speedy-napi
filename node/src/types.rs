#[napi(object)]
#[derive(Debug)]
pub struct TransformOutput {
  pub code: String,
  pub map: Option<String>,
}

#[napi(object)]
#[derive(Debug)]
pub struct TransformConfig {
  pub react_runtime: Option<bool>,
  pub babel_import: Option<Vec<BabelImportConfig>>,
}

#[napi(object)]
#[derive(Debug)]
pub struct BabelImportConfig {
  pub from_source: String,
  pub replace_css: Option<RepalceCssConfig>,
  pub replace_js: Option<RepalceSpecConfig>,
}

#[napi(object)]
#[derive(Debug)]
pub struct RepalceSpecConfig {
  pub replace_expr: String,
  pub ignore_es_component: Option<Vec<String>>,
  pub lower: Option<bool>,
}

#[napi(object)]
#[derive(Debug)]
pub struct RepalceCssConfig {
  pub ignore_style_component: Option<Vec<String>>,
  pub replace_expr: String,
  pub lower: Option<bool>,
}
