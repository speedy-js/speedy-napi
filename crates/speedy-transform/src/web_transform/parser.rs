use crate::web_transform::babel_import::transform_style;
use crate::web_transform::proxy::{Env, ExtraInfo, TransformConfig};
use crate::web_transform::react::transform_prefix_react;
use crate::web_transform::remove_effect::remove_call;

use swc::config::SourceMapsConfig;
use swc::{Compiler, TransformOutput};
use swc_common::input::StringInput;
use swc_common::sync::Lrc;
use swc_common::{FileName, SourceMap};
use swc_ecma_ast::{EsVersion, Module};
use swc_ecma_parser::lexer::Lexer;
use swc_ecma_parser::{Parser, Syntax, TsConfig};

pub fn transform_module(module: &mut Module, config: &TransformConfig, extra: &ExtraInfo) {
  transform_style(module, config, extra);
  transform_prefix_react(module, config);
  remove_call(module, config, extra);
}

pub fn transform(
  env: Env,
  code: &str,
  config: TransformConfig,
  filename: Option<String>,
  target: Option<String>,
) -> Result<TransformOutput, std::string::String> {
  let source_filename = filename.unwrap_or_else(|| "test.js".to_string());
  let cm: Lrc<SourceMap> = Default::default();
  let fm = cm.new_source_file(FileName::Custom(source_filename.clone()), code.into());
  let compiler = Compiler::new(cm);

  #[cfg(not(target_arch = "wasm32"))]
  let tsx_parse = config.tsx.unwrap_or(true);
  // wasm plugin parse option is in js side, not in rust side
  #[cfg(target_arch = "wasm32")]
  let tsx_parse = true;

  let lexer = Lexer::new(
    // We want to parse ecmascript
    Syntax::Typescript(TsConfig {
      tsx: tsx_parse,
      decorators: true,
      dts: false,
      no_early_errors: false,
    }),
    // EsVersion defaults to es5
    EsVersion::Es2016,
    StringInput::from(&*fm),
    None,
  );

  let mut parser = Parser::new_from(lexer);

  let list_error = parser.take_errors();
  if list_error.iter().len() > 0 {
    let err_msg = list_error
      .iter()
      .map(|err| err.kind().msg())
      .collect::<Vec<_>>()
      .join("");
    return Err(err_msg);
  }

  let module_result = parser.parse_module();
  if module_result.is_err() {
    return Err(module_result.err().unwrap().into_kind().msg().to_string());
  }
  let mut module = module_result.unwrap();

  #[cfg(not(target_arch = "wasm32"))]
  transform_module(
    &mut module,
    &config,
    &ExtraInfo {
      env: &env,
      compiler: &compiler,
    },
  );

  let target_ref = target.unwrap_or_else(|| "".to_string());
  let swc_target = match target_ref.as_str() {
    "" => EsVersion::Es5,
    "ES5" => EsVersion::Es5,
    "ES6" => EsVersion::Es2015,
    _ => EsVersion::Es2020,
  };

  compiler
    .print(
      &module,
      Some(source_filename.as_str()),
      None,
      false,
      swc_target,
      SourceMapsConfig::Bool(true),
      &Default::default(),
      None,
      false,
      None,
      false,
      false,
    )
    .map_err(|err| err.to_string())
}
