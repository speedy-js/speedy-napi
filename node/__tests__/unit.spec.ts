import assert from 'assert';

import {parse} from '@babel/parser';
import traverse from '@babel/traverse';
import generate from '@babel/generator';
import {Program} from '@babel/types';
import {transform} from '../lib';
import { SourceMapConsumer } from 'source-map';
import * as process from "process";

describe('speedy_napi_cases', function speedyTest() {
    it('babel_import_transform should track type correctly', async () => {
        const code = `
import { InputProps, Button } from "antd";

{
    let InputProps = 1;
    console.log(InputProps);
}

function App(props: InputProps) {}
`;

        let target_code = `
import { InputProps } from "antd";
{
    let InputProps = 1;
    console.log(InputProps);
}

function App(props: InputProps) {}
`;
        const napi_res = transform.transformBabelImport(code, {
            babelImport: [
                {
                    fromSource: 'antd',
                    replaceCss: {
                        replaceExpr: (ident: string) => {
                            return `antd/es/${ident}/style/index.css`;
                        },
                        lower: true,
                        ignoreStyleComponent: undefined,
                        camel2DashComponentName: true
                    },
                    replaceJs: {
                        replaceExpr: (ident: string) => {
                            return `antd/es/${ident}/index.js`;
                        },
                        lower: true,
                        ignoreEsComponent: undefined,
                        camel2DashComponentName: true
                    },
                },
            ]
        });

        assert.equal(
            target_code.replace(/\ +/g, '').replace(/[\r\n]/g, ''),
            napi_res.code.replace(/\ +/g, '').replace(/[\r\n]/g, '')
        );
    });

    it('babel_import_transfrom with camel2DashComponentName true', async () => {
        const code = `
import React from "react";
import ReactDOM from "react-dom";
import { Input, AutoComplete } from "antd";
import Child from "./component/Child";
import { Button as AntButton } from "antd";

class Page extends React.Component<any,any> {
    render() {
        return (
            <div className={"test"}>
                <div>Page</div>
                <Child/>
                <AntButton>click me</AntButton>
                <Input/>
                <AutoComplete />
            </div>
        );
    }
}

ReactDOM.render(<Page/>, document.getElementById("root"));
`;

        let target_code = `
import "antd/es/button/style/index.css";
import "antd/es/auto-complete/style/index.css";
import "antd/es/input/style/index.css";
import AntButton from "antd/es/button/index.js";
import AutoComplete from "antd/es/auto-complete/index.js";
import Input from "antd/es/input/index.js";
import React from "react";
import ReactDOM from "react-dom";
import Child from "./component/Child";

class Page extends React.Component<any,any>{
    render() {
        return (
            <div className={"test"}>
                <div>Page</div>
                <Child/>
                <AntButton>click me</AntButton>
                <Input/>
                <AutoComplete />
            </div>
        );
    }
}

ReactDOM.render(<Page / >, document.getElementById("root"));
        `;
        console.time('babel_import_swc_transfrom');
        process.env["rsdebug"] = "info";
        const napi_res = transform.transformBabelImport(code, {
            reactRuntime: true,
            babelImport: [
                {
                    fromSource: 'antd',
                    replaceCss: {
                        replaceExpr: (ident: string) => {
                            return `antd/es/${ident}/style/index.css`;
                        },
                        lower: true,
                        ignoreStyleComponent: undefined,
                        camel2DashComponentName: true
                    },
                    replaceJs: {
                        replaceExpr: (ident: string) => {
                            return `antd/es/${ident}/index.js`;
                        },
                        lower: true,
                        ignoreEsComponent: undefined,
                        camel2DashComponentName: true
                    },
                },
            ]
        })
        console.timeEnd('babel_import_swc_transfrom');

        // 执行同样的 babel 操作
        console.time('babel_import_babeljs_transfrom');

        const babel_res = babel_impl_bableimport(code, 'antd', `antd/es/{}/style/index.css`);
        console.timeEnd('babel_import_babeljs_transfrom');

        assert.equal(
            target_code.replace(/\ +/g, '').replace(/[\r\n]/g, ''),
            napi_res.code.replace(/\ +/g, '').replace(/[\r\n]/g, '')
        );
    });

    it('babel_import_transfrom with transformToDefaultImport set false', async () => {
        const code = `
import React from "react";
import ReactDOM from "react-dom";
import { Input, AutoComplete } from "antd";
import Child from "./component/Child";
import { Button as AntButton } from "antd";

class Page extends React.Component<any,any> {
    render() {
        return (
            <div className={"test"}>
                <div>Page</div>
                <Child/>
                <AntButton>click me</AntButton>
                <Input/>
                <AutoComplete />
            </div>
        );
    }
}

ReactDOM.render(<Page/>, document.getElementById("root"));
`;

        let target_code = `
import "antd/es/button/style/index.css";
import "antd/es/auto-complete/style/index.css";
import "antd/es/input/style/index.css";
import { Button as AntButton } from "antd/es/button/index.js";
import { AutoComplete } from "antd/es/auto-complete/index.js";
import { Input } from "antd/es/input/index.js";
import React from "react";
import ReactDOM from "react-dom";
import Child from "./component/Child";

class Page extends React.Component<any,any>{
    render() {
        return (
            <div className={"test"}>
                <div>Page</div>
                <Child/>
                <AntButton>click me</AntButton>
                <Input/>
                <AutoComplete />
            </div>
        );
    }
}

ReactDOM.render(<Page / >, document.getElementById("root"));
        `;
        console.time('babel_import_swc_transfrom');
        process.env["rsdebug"] = "info";
        const napi_res = transform.transformBabelImport(code, {
            reactRuntime: true,
            babelImport: [
                {
                    fromSource: 'antd',
                    replaceCss: {
                        replaceExpr: (ident: string) => {
                            return `antd/es/${ident}/style/index.css`;
                        },
                        lower: true,
                        ignoreStyleComponent: undefined,
                        camel2DashComponentName: true,
                    },
                    replaceJs: {
                        replaceExpr: (ident: string) => {
                            return `antd/es/${ident}/index.js`;
                        },
                        lower: true,
                        ignoreEsComponent: undefined,
                        transformToDefaultImport: false,
                        camel2DashComponentName: true,
                    },
                },
            ]
        })
        console.timeEnd('babel_import_swc_transfrom');

        // 执行同样的 babel 操作
        console.time('babel_import_babeljs_transfrom');

        const babel_res = babel_impl_bableimport(code, 'antd', `antd/es/{}/style/index.css`);
        console.timeEnd('babel_import_babeljs_transfrom');

        assert.equal(
            target_code.replace(/\ +/g, '').replace(/[\r\n]/g, ''),
            napi_res.code.replace(/\ +/g, '').replace(/[\r\n]/g, '')
        );
    });

    it('babel_import_transfrom should tree shaking (ts_type and unused components)', async () => {
        const code = `
import React from "react";
import ReactDOM from "react-dom";
import { Input, AutoComplete, InputProps, Radio } from "antd";
import Child from "./component/Child";

type Props = InputProps;

class Page extends React.Component<Props,any> {
    render() {
        return (
            <div className={"test"}>
                <div>Page</div>
                <Input/>
                <Radio.Group />
            </div>
        );
    }
}

ReactDOM.render(<Page/>, document.getElementById("root"));
`;

        let target_code = `
import "antd/es/radio/style/index.css";
import "antd/es/input/style/index.css";
import { Radio } from "antd/es/radio/index.js";
import { Input } from "antd/es/input/index.js";
import React from "react";
import ReactDOM from "react-dom";
import { InputProps } from "antd";
import Child from "./component/Child";

type Props = InputProps;

class Page extends React.Component<Props,any>{
    render() {
        return (
            <div className={"test"}>
                <div>Page</div>
                <Input/>
                <Radio.Group />
            </div>
        );
    }
}

ReactDOM.render(<Page / >, document.getElementById("root"));
        `;
        console.time('babel_import_swc_transfrom');
        process.env["rsdebug"] = "info";
        const napi_res = transform.transformBabelImport(code, {
            reactRuntime: true,
            babelImport: [
                {
                    fromSource: 'antd',
                    replaceCss: {
                        replaceExpr: (ident: string) => {
                            return `antd/es/${ident}/style/index.css`;
                        },
                        lower: true,
                        ignoreStyleComponent: undefined,
                        camel2DashComponentName: true,
                    },
                    replaceJs: {
                        replaceExpr: (ident: string) => {
                            return `antd/es/${ident}/index.js`;
                        },
                        lower: true,
                        ignoreEsComponent: undefined,
                        transformToDefaultImport: false,
                        camel2DashComponentName: true,
                    },
                },
            ]
        })
        console.timeEnd('babel_import_swc_transfrom');

        // 执行同样的 babel 操作
        console.time('babel_import_babeljs_transfrom');

        const babel_res = babel_impl_bableimport(code, 'antd', `antd/es/{}/style/index.css`);
        console.timeEnd('babel_import_babeljs_transfrom');

        assert.equal(
            target_code.replace(/\ +/g, '').replace(/[\r\n]/g, ''),
            napi_res.code.replace(/\ +/g, '').replace(/[\r\n]/g, '')
        );
    });

    it('babel_import_transfrom should track components ref correctly', async () => {
        const code = `
import React from "react";
import ReactDOM from "react-dom";
import { Radio, List } from "antd";

const Item = List.Item; // ref List

class Page extends React.Component<InputProps,any> {
    render() {
        return (
            <div className={"test"}>
                {/* ref Radio */}
                <Radio.RadioGroup.RadioItem />
                <Item />
            </div>
        );
    }
}

ReactDOM.render(<Page/>, document.getElementById("root"));
`;

        let target_code = `
import "antd/es/list/style/index.css";
import "antd/es/radio/style/index.css";
import { List } from "antd/es/list/index.js";
import { Radio } from "antd/es/radio/index.js";
import React from "react";
import ReactDOM from "react-dom";

const Item = List.Item;

class Page extends React.Component<InputProps,any>{
    render() {
        return (
            <div className={"test"}>
                {}
                <Radio.RadioGroup.RadioItem />
                <Item />
            </div>
        );
    }
}

ReactDOM.render(<Page / >, document.getElementById("root"));
        `;
        console.time('babel_import_swc_transfrom');
        process.env["rsdebug"] = "info";
        const napi_res = transform.transformBabelImport(code, {
            reactRuntime: true,
            babelImport: [
                {
                    fromSource: 'antd',
                    replaceCss: {
                        replaceExpr: (ident: string) => {
                            return `antd/es/${ident}/style/index.css`;
                        },
                        lower: true,
                        ignoreStyleComponent: undefined,
                        camel2DashComponentName: true,
                    },
                    replaceJs: {
                        replaceExpr: (ident: string) => {
                            return `antd/es/${ident}/index.js`;
                        },
                        lower: true,
                        ignoreEsComponent: undefined,
                        transformToDefaultImport: false,
                        camel2DashComponentName: true,
                    },
                },
            ]
        })
        console.timeEnd('babel_import_swc_transfrom');

        // 执行同样的 babel 操作
        console.time('babel_import_babeljs_transfrom');

        const babel_res = babel_impl_bableimport(code, 'antd', `antd/es/{}/style/index.css`);
        console.timeEnd('babel_import_babeljs_transfrom');

        assert.equal(
            target_code.replace(/\ +/g, '').replace(/[\r\n]/g, ''),
            napi_res.code.replace(/\ +/g, '').replace(/[\r\n]/g, '')
        );
    });

    it('remove_call_transform should work with simple case', async () => {
        let code = `
import React from 'react';
import ReactDOM from "react-dom";
import { useEffect } from 'react';

function App() {
    const [num, setNum] = React.useState(1);
    React.useState(2);
    
    React.useEffect(() => {
        setNum(2);
    }, []);

    useEffect(() => {
        setNum(3);
    }, []);

    return <div>{num}</div>;
}
ReactDOM.render(<Page/>, document.getElementById("root"));
`;

        let target_code = `
import React from "react";
import ReactDOM from "react-dom";
import { useEffect } from "react";

function App() {
    const [num, setNum] = React.useState(1);
    React.useState(2);

    return <div>{num}</div>;
}

ReactDOM.render(<Page/>, document.getElementById("root"));
`;

        const napi_res = transform.transformBabelImport(code, {
            removeUseEffect: true,
        })

        assert.equal(
            target_code.replace(/\ +/g, '').replace(/[\r\n]/g, ''),
            napi_res.code.replace(/\ +/g, '').replace(/[\r\n]/g, '')
        );
    })

    it('remove_call_transform should work with complex case', async () => {
        let code = `
import Recta from 'react';
import ReactDOM from "react-dom";
import { useEffect as effectUse } from 'react';

function useEffect() {
    console.log("not delete");
}

{
    useEffect();
}

function App() {
    const [num, setNum] = Recta.useState(1);
    Recta.useState(1);
    
    Recta.useEffect(() => {
        setNum(2);
    }, []);

    effectUse(() => {
        setNum(3);
    }, []);

    {
        effectUse(() => {
            setNum(4);
        }, []);
    }

    {
        const useEffect = () => 2;
        const effectUse = () => 1;
        useEffect();
        effectUse();
    }

    return <div>{num}</div>;
}
ReactDOM.render(<Page/>, document.getElementById("root"));
`;

        let target_code = `
import Recta from "react";
import ReactDOM from "react-dom";
import { useEffect as effectUse } from "react";

function useEffect() {
    console.log("not delete");
}

{
    useEffect();
}

function App() {
    const [num, setNum] = Recta.useState(1);
    Recta.useState(1);

    {}

    {
        const useEffect = () => 2;
        const effectUse = () => 1;
        useEffect();
        effectUse();
    }

    return <div>{num}</div>;
}

ReactDOM.render(<Page/>, document.getElementById("root"));
`;

        const napi_res = transform.transformBabelImport(code, {
            removeUseEffect: true,
        })

        assert.equal(
            target_code.replace(/\ +/g, '').replace(/[\r\n]/g, ''),
            napi_res.code.replace(/\ +/g, '').replace(/[\r\n]/g, '')
        );
    })
    
    it(`remove_call_transform should work correctly among scope`, async () => {
        // https://github.com/speedy-js/speedy-native/pull/27#issuecomment-1195278186
        let code = `
import { useEffect } from 'react';

{
    const useEffect = () => {}
    useEffect()
}

function App() {
    useEffect()
}`;

        let target_code = `
import { useEffect } from "react";
{
    const useEffect = ()=>{};
    useEffect();
}
function App() {}
`;

        const napi_res = transform.transformBabelImport(code, {
            removeUseEffect: true,
        })

        assert.equal(
            target_code.replace(/\ +/g, '').replace(/[\r\n]/g, ''),
            napi_res.code.replace(/\ +/g, '').replace(/[\r\n]/g, '')
        );
    })

    it(`remove_call source map test`, async () => {
        let code = `
import React from "react";
import ReactDOM from "react-dom";
import { useEffect } from 'react';

function App() {
    const [num, setNum] = React.useState(1);
    
    React.useEffect(() => {
        setNum(2);
    }, []);

    useEffect(() => {
        setNum(3);
    }, []);

    return (
        <div>{num}</div>
    );
}
ReactDOM.render(<Page/>, document.getElementById("root"));
`;

        let target_code = `import React from "react";
import ReactDOM from "react-dom";
import { useEffect } from "react";
function App() {
    const [num, setNum] = React.useState(1);
    return (<div >{num}</div>);
}
ReactDOM.render(<Page />, document.getElementById("root"));
`;

        const napi_res = transform.transformBabelImport(code, {
            removeUseEffect: true
        });

        const consumer = await new SourceMapConsumer(napi_res.map as any);

        const position1 = consumer.originalPositionFor({
            line: 5,
            column: 4
        })
        const position2 = consumer.originalPositionFor({
            line: 6,
            column: 12
        })
        const position3 = consumer.originalPositionFor({
            line: 8,
            column: 1
        })

        assert.equal(napi_res.code, target_code);
        assert.equal(position1.line, 7);
        assert.equal(position1.column, 4);
        assert.equal(position2.line, 17);
        assert.equal(position2.column, 4);
        assert.equal(position3.line, 21);
        assert.equal(position3.column, 0);
    })
});

/*
 * babel 同样实现 性能比对函数
 */
function babel_impl_bableimport(code: string, lib: string, expr: string) {
    /** 解析源码AST树 */
    const ast: any = parse(code, {
        sourceType: 'module',
        sourceFilename: undefined,
        plugins: [
            'typescript',
            'jsx',
            'decorators-legacy',
            'classProperties',
            'bigInt',
            'importMeta',
            'optionalChaining',
            'nullishCoalescingOperator',
            'importMeta',
            'optionalCatchBinding',
        ],
    });

    const change_ast = (ast: any) => {
        let pro: Program;
        traverse(ast, {
            enter(path) {
                if (path.isProgram()) {
                    pro = path.node;
                }
                const matchsource = (source: string) => {
                    return lib == source;
                };
                if (path.isImportDeclaration() && matchsource(path.node.source.value)) {
                    const origin_names = path.node.specifiers.map((p) => {
                        return expr.replace('{}', p.local.name);
                    });
                    origin_names.forEach((style_source) => {
                        pro.body.unshift({
                            type: 'ImportDeclaration',
                            specifiers: [],
                            source: {
                                type: 'StringLiteral',
                                extra: {
                                    rawValue: style_source,
                                    raw: `'${style_source}'`,
                                },
                                value: style_source,
                            },
                        } as any);
                    });
                }
                return;
            },
        });
    };

    change_ast(ast);
    const res = generate(ast, {sourceMaps: true, sourceFileName: 'test.js'}, code);
    return res;
}
