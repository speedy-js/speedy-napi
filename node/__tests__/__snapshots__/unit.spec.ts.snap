// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`speedy-napi: babel-import babel_import_transform should track type correctly 1`] = `
"import { InputProps } from \\"antd\\";
{
    let InputProps = 1;
    console.log(InputProps);
}function App(props: InputProps) {}
"
`;

exports[`speedy-napi: babel-import babel_import_transfrom should track components ref correctly 1`] = `
"import \\"antd/es/list/style/index.css\\";
import \\"antd/es/radio/style/index.css\\";
import { List } from \\"antd/es/list/index.js\\";
import { Radio } from \\"antd/es/radio/index.js\\";
import React from \\"react\\";
import ReactDOM from \\"react-dom\\";
const Item = List.Item;
class Page extends React.Component<InputProps, any> {
    render() {
        return (<div className={\\"test\\"}>

                {}

                <Radio.RadioGroup.RadioItem />

                <Item />

            </div>);
    }
}
ReactDOM.render(<Page />, document.getElementById(\\"root\\"));
"
`;

exports[`speedy-napi: babel-import babel_import_transfrom should tree shaking (ts_type and unused components) 1`] = `
"import \\"antd/es/radio/style/index.css\\";
import \\"antd/es/input/style/index.css\\";
import { Radio } from \\"antd/es/radio/index.js\\";
import { Input } from \\"antd/es/input/index.js\\";
import React from \\"react\\";
import ReactDOM from \\"react-dom\\";
import { InputProps } from \\"antd\\";
import Child from \\"./component/Child\\";
type Props = InputProps;
class Page extends React.Component<Props, any> {
    render() {
        return (<div className={\\"test\\"}>

                <div >Page</div>

                <Input />

                <Radio.Group />

            </div>);
    }
}
ReactDOM.render(<Page />, document.getElementById(\\"root\\"));
"
`;

exports[`speedy-napi: babel-import babel_import_transfrom with camel2DashComponentName true 1`] = `
"import \\"antd/es/button/style/index.css\\";
import \\"antd/es/auto-complete/style/index.css\\";
import \\"antd/es/input/style/index.css\\";
import AntButton from \\"antd/es/button/index.js\\";
import AutoComplete from \\"antd/es/auto-complete/index.js\\";
import Input from \\"antd/es/input/index.js\\";
import React from \\"react\\";
import ReactDOM from \\"react-dom\\";
import Child from \\"./component/Child\\";
class Page extends React.Component<any, any> {
    render() {
        return (<div className={\\"test\\"}>

                <div >Page</div>

                <Child />

                <AntButton >click me</AntButton>

                <Input />

                <AutoComplete />

            </div>);
    }
}
ReactDOM.render(<Page />, document.getElementById(\\"root\\"));
"
`;

exports[`speedy-napi: babel-import babel_import_transfrom with transformToDefaultImport set false 1`] = `
"import \\"antd/es/button/style/index.css\\";
import \\"antd/es/auto-complete/style/index.css\\";
import \\"antd/es/input/style/index.css\\";
import { Button as AntButton } from \\"antd/es/button/index.js\\";
import { AutoComplete } from \\"antd/es/auto-complete/index.js\\";
import { Input } from \\"antd/es/input/index.js\\";
import React from \\"react\\";
import ReactDOM from \\"react-dom\\";
import Child from \\"./component/Child\\";
class Page extends React.Component<any, any> {
    render() {
        return (<div className={\\"test\\"}>

                <div >Page</div>

                <Child />

                <AntButton >click me</AntButton>

                <Input />

                <AutoComplete />

            </div>);
    }
}
ReactDOM.render(<Page />, document.getElementById(\\"root\\"));
"
`;

exports[`speedy-napi: remove call remove_call source map test 1`] = `
"import React from \\"react\\";
import ReactDOM from \\"react-dom\\";
import { useEffect } from \\"react\\";
function App() {
    const [num, setNum] = React.useState(1);
    return (<div >{num}</div>);
}
ReactDOM.render(<Page />, document.getElementById(\\"root\\"));
"
`;

exports[`speedy-napi: remove call remove_call source map test 2`] = `7`;

exports[`speedy-napi: remove call remove_call source map test 3`] = `4`;

exports[`speedy-napi: remove call remove_call source map test 4`] = `17`;

exports[`speedy-napi: remove call remove_call source map test 5`] = `4`;

exports[`speedy-napi: remove call remove_call source map test 6`] = `21`;

exports[`speedy-napi: remove call remove_call source map test 7`] = `0`;

exports[`speedy-napi: remove call remove_call_transform should work correctly among scope 1`] = `
"import { useEffect } from \\"react\\";
{
    const useEffect = ()=>{};
    useEffect();
}function App() {}
"
`;

exports[`speedy-napi: remove call remove_call_transform should work with complex case 1`] = `
"import Recta from \\"react\\";
import ReactDOM from \\"react-dom\\";
import { useEffect as effectUse } from \\"react\\";
function useEffect() {
    console.log(\\"not delete\\");
}
{
    useEffect();
}function App() {
    const [num, setNum] = Recta.useState(1);
    Recta.useState(1);
    {}
    {
        const useEffect = ()=>2;
        const effectUse = ()=>1;
        useEffect();
        effectUse();
    }
    return <div >{num}</div>;
}
ReactDOM.render(<Page />, document.getElementById(\\"root\\"));
"
`;

exports[`speedy-napi: remove call remove_call_transform should work with import * as 1`] = `
"import * as React from \\"react\\";
import ReactDOM from \\"react-dom\\";
import { useEffect } from \\"react\\";
function App() {
    const [num, setNum] = React.useState(1);
    React.useState(2);
    return <div >{num}</div>;
}
ReactDOM.render(<Page />, document.getElementById(\\"root\\"));
"
`;

exports[`speedy-napi: remove call remove_call_transform should work with multi import 1`] = `
"import * as React from \\"react\\";
import ReactDOM from \\"react-dom\\";
import ReactDefault, { useEffect } from \\"react\\";
import { useEffect as useEffect2 } from \\"react\\";
import * as AnotherReact from \\"react\\";
function App() {
    const [num, setNum] = React.useState(1);
    React.useState(2);
    return <div >{num}</div>;
}
ReactDOM.render(<Page />, document.getElementById(\\"root\\"));
"
`;

exports[`speedy-napi: remove call remove_call_transform should work with simple case 1`] = `
"import React from \\"react\\";
import ReactDOM from \\"react-dom\\";
import { useEffect } from \\"react\\";
function App() {
    const [num, setNum] = React.useState(1);
    React.useState(2);
    return <div >{num}</div>;
}
ReactDOM.render(<Page />, document.getElementById(\\"root\\"));
"
`;

exports[`speedy-napi: ts parse config can parse ts only syntax 1`] = `
"import { useEffect, useState } from \\"react\\";
function useCount() {
    const [count, setCount] = useState(0);
    return [
        count,
        setCount
    ];
}
const useName = <[]>useCount();
"
`;